lib/core/src/wallet.rs

```diff
@@ -0,0 +1,145 @@
+        let is_mainnet = config.network == Network::Mainnet;
+        let lwk_signer = SwSigner::new(&mnemonic, is_mainnet)?;
+        let descriptor = LiquidOnchainWallet::get_descriptor(&lwk_signer, config.network)?;
+        let elements_network: ElementsNetwork = config.network.into();
+
+        let lwk_persister =
+            FsPersister::new(config.working_dir.clone(), elements_network, &descriptor)?;
+        let wollet = Wollet::new(elements_network, lwk_persister, descriptor)?;
+        Ok(Self {
+            wallet: Arc::new(Mutex::new(wollet)),
+            lwk_signer,
+            config,
+        })
+    }
+
+    fn get_descriptor(
+        signer: &SwSigner,
+        network: Network,
+    ) -> Result<WolletDescriptor, PaymentError> {
+        let is_mainnet = network == Network::Mainnet;
+        let descriptor_str = singlesig_desc(
+            signer,
+            Singlesig::Wpkh,
+            lwk_common::DescriptorBlindingKey::Slip77,
+            is_mainnet,
+        )
+        .map_err(|e| anyhow!("Invalid descriptor: {e}"))?;
+        Ok(descriptor_str.parse()?)
+    }
+}
+
+#[async_trait]
+impl OnchainWallet for LiquidOnchainWallet {
+    /// List all transactions in the wallet
+    async fn transactions(&self) -> Result<Vec<WalletTx>, PaymentError> {
+        let wallet = self.wallet.lock().await;
+        wallet.transactions().map_err(|e| PaymentError::Generic {
+            err: format!("Failed to fetch wallet transactions: {e:?}"),
+        })
+    }
+
+    /// Build a transaction to send funds to a recipient
+    async fn build_tx(
+        &self,
+        fee_rate: Option<f32>,
+        recipient_address: &str,
+        amount_sat: u64,
+    ) -> Result<Transaction, PaymentError> {
+        let lwk_wollet = self.wallet.lock().await;
+        let mut pset = lwk_wollet::TxBuilder::new(self.config.network.into())
+            .add_lbtc_recipient(
+                &ElementsAddress::from_str(recipient_address).map_err(|e| {
+                    PaymentError::Generic {
+                        err: format!(
+                      "Recipient address {recipient_address} is not a valid ElementsAddress: {e:?}"
+                  ),
+                    }
+                })?,
+                amount_sat,
+            )?
+            .fee_rate(fee_rate)
+            .finish(&lwk_wollet)?;
+        let signer = AnySigner::Software(self.lwk_signer.clone());
+        signer.sign(&mut pset)?;
+        Ok(lwk_wollet.finalize(&mut pset)?)
+    }
+
+    /// Get the next unused address in the wallet
+    async fn next_unused_address(&self) -> Result<Address, PaymentError> {
+        Ok(self.wallet.lock().await.address(None)?.address().clone())
+    }
+
+    /// Get the current tip of the blockchain the wallet is aware of
+    async fn tip(&self) -> Tip {
+        self.wallet.lock().await.tip()
+    }
+
+    /// Get the public key of the wallet
+    async fn pubkey(&self) -> String {
+        self.lwk_signer.xpub().to_string()
```

Comment:
Seems like the CI is failing because of this, should it be?
```suggestion
        self.lwk_signer.xpub().public_key.to_string()
```

Reply:
also I think the function doesn't really need to be `async`